### PR TITLE
fix policy cache missing

### PR DESCRIPTION
**Pull request by kevindiu**

### CODE REVIEW

File: policy/daemon.go [hunk 1]
```diff
@@ -330,7 +310,7 @@ func (p *policyd) fetchPolicy(ctx context.Context, domain string) (*SignedPolicy
 	return sp, true, nil
 }
 
-func simplifyAndCachePolicy(ctx context.Context, rp gache.Gache, sp *SignedPolicy) error {
+func simplifyAndCachePolicy(ctx context.Context, dom string, sp *SignedPolicy) (map[string][]*Assertion, error) {
```
Comment (WindzCUHK) [inline policy/daemon.go]:
```suggestion
func simplifyPolicy(ctx context.Context, dom string, sp *SignedPolicy) (map[string][]*Assertion, error) {
```

File: policy/daemon.go [hunk 2]
```diff
@@ -343,6 +323,12 @@ func simplifyAndCachePolicy(ctx context.Context, rp gache.Gache, sp *SignedPolic
 				case <-ctx.Done():
 					return ctx.Err()
 				default:
+					if strings.SplitN(ass.Role, ":role.", 2)[0] != dom {
+						glg.Errorf("role is not matches with the domain, role: %v, domain: %v", ass.Role, dom)
+						continue
+					}
+
+					// remove duplication of `role+action+reource' assertion
```
Comment (WindzCUHK) [inline policy/daemon.go]:
```suggestion
					// remove duplication of `role+action+resource' assertion
```

File: policy/daemon.go [hunk 3]
```diff
@@ -343,6 +323,12 @@ func simplifyAndCachePolicy(ctx context.Context, rp gache.Gache, sp *SignedPolic
 				case <-ctx.Done():
 					return ctx.Err()
 				default:
+					if strings.SplitN(ass.Role, ":role.", 2)[0] != dom {
+						glg.Errorf("role is not matches with the domain, role: %v, domain: %v", ass.Role, dom)
```
Comment (WindzCUHK) [inline policy/daemon.go]:
```suggestion
						glg.Debugf("role is not matches with the domain, role: %v, domain: %v", ass.Role, dom)
```

File: policy/daemon.go [hunk 4]
```diff
@@ -358,37 +344,31 @@ func simplifyAndCachePolicy(ctx context.Context, rp gache.Gache, sp *SignedPolic
...
 
+	rp := make(map[string][]*Assertion)
 	// cache
 	var retErr error
 	assm.Range(func(k interface{}, val interface{}) bool {
 		ass := val.(*util.Assertion)
 		a, err := NewAssertion(ass.Action, ass.Resource, ass.Effect)
 		if err != nil {
-			glg.Debugf("error adding assertion to the cache, err: %v", err)
+			glg.Errorf("error adding assertion to the cache, assertion: %+v,  err: %v", ass, err)
```
Comment (WindzCUHK) [inline policy/daemon.go]:
```suggestion
			glg.Errorf("error creating policy.Assertion, assertion: %+v,  err: %v", ass, err)
```

File: policy/daemon.go [hunk 5]
```diff
@@ -100,6 +100,12 @@ func (p *policyd) Start(ctx context.Context) <-chan error {
 		defer close(fch)
 		defer close(ech)
 
+		p.rolePolicies.StartExpired(ctx, p.policyExpiredDuration).
+			EnableExpiredHook().
+			SetExpiredHook(func(ctx context.Context, key string) {
+				//key = <domain>:role.<role>
+				p.fetchAndCachePolicy(ctx, strings.Split(key, ":role.")[0])
+			})
```
Comment (WindzCUHK) [inline policy/daemon.go]: 1. ExpiredHook is called after delete, checking policy between time gap with cause error
2. retry on ExpiredHook fail?
3. fetchAndCachePolicy() error is not print inside the ExpiredHook

File: policy/daemon.go [hunk 6]
```diff
@@ -168,93 +172,69 @@ func (p *policyd) Update(ctx context.Context) error {
...
-						case <-cctx.Done():
-							ch <- cctx.Err()
-							return
-						default:
+				if roleAsss, ok := p.rolePolicies.Get(domain); ok {
+					if asss, ok := roleAsss.(map[string][]*Assertion)[r]; ok {
+						for _, ass := range asss {
+							glg.Debugf("Checking policy domain: %s, role: %v, action: %s, resource: %s, assertion: %v", domain, roles, action, resource, ass)
 							if strings.EqualFold(ass.ResourceDomain, domain) && ass.Reg.MatchString(strings.ToLower(action+"-"+resource)) {
 								ch <- ass.Effect
```
Comment (WindzCUHK) [inline policy/daemon.go]: I think we will have deadlock here, when multiple deny policy match.
1. ech size = 1
2. read from ech only 1 times
3. when, write more than 2, the go func cannot return, wg.Wait() will block forever

better to keep the old logic first

File: policy/daemon.go [hunk 7]
```diff
@@ -168,93 +172,69 @@ func (p *policyd) Update(ctx context.Context) error {
...
 // CheckPolicy checks the specified request has privilege to access the resources or not.
 // If return is nil then the request is allowed, otherwise the request is rejected.
 func (p *policyd) CheckPolicy(ctx context.Context, domain string, roles []string, action, resource string) error {
-	ech := make(chan error, 1)
-	cctx, cancel := context.WithCancel(ctx)
-	defer cancel()
+	ch := make(chan error, 1)
 
 	go func() {
-		defer close(ech)
```
Comment (WindzCUHK) [inline policy/daemon.go]: better to close the channel explicitly

File: policy/daemon.go (hunk 5 above)
Comment (WindzCUHK) [reply policy/daemon.go]: refer to: https://github.com/yahoojapan/athenz-authorizer/issues/29

Comment (WindzCUHK) [reply policy/daemon.go]: refer to: https://github.com/yahoojapan/athenz-authorizer/issues/28

File: policy/daemon.go (hunk 6 above)
Comment (WindzCUHK) [reply policy/daemon.go]: refer to: https://github.com/yahoojapan/athenz-authorizer/issues/30

**Comment by kpango** [comment]
closed by #27

**Pull request closed without merging**